### PR TITLE
SE-0135

### PR DESCRIPTION
**Pull request by ddunbar**
This adds support for SE-0135.

It isn't particularly pretty, but it has minimal impact on the current code which is important at this stage.

### CODE REVIEW

**Comment by ddunbar** [comment]
I still need to do more testing on this on real world packages, but I believe it implements the feature completely.

/cc @aciidb0mb3r, @abertelrud review appreciated.

File: Sources/Utility/Git.swift [hunk 1]
```diff
@@ -59,14 +59,34 @@ public class Git {
             let out = (try? Git.runPopen([Git.tool, "-C", repo.path.asString, "tag", "-l"])) ?? ""
             let tags = out.characters.split(separator: "\n").map{ String($0) }
 
-            // First try the plain init.
+            // First, check if we need to restrict the tag set to version-specific tags.
             var knownVersions: [Version: String] = [:]
+            for versionSpecificKey in Versioning.currentVersionSpecificKeys {
+                for tag in tags {
+                    if tag.hasSuffix(versionSpecificKey) {
```
Comment (aciidb0mb3r) [inline Sources/Utility/Git.swift]: nit: `for tag in tags where tag.hasSuffix(versionSpecificKey) {`

File: Sources/Utility/Git.swift [hunk 2]
```diff
@@ -59,14 +59,34 @@ public class Git {
...
+                        let specifier = String(tag.characters.dropLast(versionSpecificKey.characters.count))
+                        if let version = Version(specifier) ?? Version.vprefix(specifier) {
```
Comment (aciidb0mb3r) [inline Sources/Utility/Git.swift]: unrelated but is there a reason for vprefix method being separate?

Comment (ddunbar) [inline Sources/Utility/Git.swift]: Just historical... the old code was maybe intentionally trying to ignore any v-prefixed versions if no non-prefixed versions are found, but I think that is a misfeature.

File: Sources/Utility/Versioning.swift [hunk 3]
```diff
@@ -27,9 +27,14 @@ public struct SwiftVersion {
     /// Build information, as an unstructured string.
     public var buildIdentifier: String?
 
+    /// The major version number
```
Comment (aciidb0mb3r) [inline Sources/Utility/Versioning.swift]: nit: punctuation at the sentence end

File: Sources/Utility/Versioning.swift [hunk 4]
```diff
@@ -47,6 +52,18 @@ public struct SwiftVersion {
...
     }
+
+    /// The list of version specific identifiers to search when attempting to
+    /// load version specific package or version information, in order of
+    /// preference.
+    public var versionSpecificKeys: [String] {
+        return [
+            "@swift-\(major).\(minor).\(patch)",
+            "@swift-\(major).\(minor)",
+            "@swift-\(major)"
```
Comment (aciidb0mb3r) [inline Sources/Utility/Versioning.swift]: 👍

**Comment by aciidb0mb3r** [comment]
LGTM

File: Sources/Utility/Versioning.swift (hunk 3 above)
Comment (ddunbar) [inline Sources/Utility/Versioning.swift]: faf1e67

File: Sources/Utility/Git.swift (hunk 1 above)
Comment (ddunbar) [inline Sources/Utility/Git.swift]: faf1e67

**Comment by ddunbar** [comment]
@swift-ci please test

**Comment by ddunbar** [comment]
Found a couple loose ends, the `Get` module needed to be taught about this, as did conventions (to ignore version-specific manifests).

@swift-ci please test

**Comment by ddunbar** [comment]
CI seems unusable at the moment, testing locally.

**Pull request merged**